872 ui change for surface geometry searching

juzen2003 opened the pull request:
- Fixes #872, Related to this pull: https://github.com/SETI/pds-opus/pull/956
- Were any Django, import pipeline, table_schema, or dictionary files modified? Y
  - Database used: opus3_test_200107 (obs_surface_geometry_target.json is modified)
  - All Django tests pass: Y (some tests are modified and added)
- Were any JavaScript or CSS files modified? Y
  - JSHINT run on all affected files: Y
  - Tested on Chrome, Firefox, Safari, and iOS: Y
    (Remember to test all browser sizes)
  - Tested for race conditions (with delayed API calls if applicable): Y

Description of changes:
- Fixed the issue of blank rows in selected metadata list when switching between "Pan" and "Pandora" in surfacegeo target name.
- Fixed the issue of all radio buttons being de-selected when surfacegeo target widget is dragged.
- Remove "Surface Geometry Target Name" from the "Current Search Terms" part of select metadata.
- Make sure new input can be added/removed after selecting a new surfacegeo target.

Known problems: (Fixed)
- From https://github.com/SETI/pds-opus/pull/956
    - If you click on radio buttons rapidly, you can get the selected metadata list to screw up. This is especially true with "Pan" and "Pandora".
    - The "Surface Geometry Target Name" field should never show up in the "Current Search Terms" part of select metadata.
    - It's possible to-deselect all radio buttons by clicking just to the left of the selected button. 
- If the user switches to a new surfacegeo target, the add input button will not work in related SURFACEGEO widgets after DOM attributes are updated in place.

juzen2003 committed:
Add each surfacegeo target slug into the data-slug attribute of the corresponding
surfacegeo radio input.
- In widgets.js (line 755-764), create getSurfacegeoTargetSlug to convert
  surfacegeo pretty name to surfacegeo slug.
- In widgets.js (line 741-748), put each surfacegeo target slug into the
  data-slug attribute of the corresponding radio input.
- In import_util.py (line 265-266), add a comment as a reminder to change
  getSurfacegeoTargetSlug in widgets.js whenever slug_name_for_sfc_target
  function is changed.

juzen2003 committed:
Create oldSurfacegeoTarget to store the current surfacegeo target slug.
This will be used to detect the old surfacegeo related slugs for swapping
with the new surfacegeo related slugs.
- In hash.js (line 644-650), init opus.oldSurfacegeoTarget in initFromHash
  if the reloading URL has surfacegeometrytargetname.
- In opus.js (line 146), declare oldSurfacegeoTarget to detect the old
  surfacegeo related slugs.
- In utils.js (line 114-123), move getSurfacegeoTargetSlug to utils.js.
- In widgets.js and search.js, add debugging console.log, will remove them
  later.
- In import_util.py (line 266), update comments.

juzen2003 committed:
Update surfacegeo attributes in DOMs when surfacegeo target name is changed.
Also properly update selections, extras, and widgets column to have the
correct URL updated.
- In widget.html (line 14), add class="op-widget-title" to widget title.
  This is to update the widget title in place more easily.
- In widgets.js (line 98-198), modify the singlechoice change event handler
  to properly update related surfacegeo attributes in DOMs, and also update
  opus.selections, opus.extras, and opus.prefs.widgets to have the correct
  URL updated.
- In widgets.js (line 228-240), create updateSURFACEGEOattrInPlace to update
  surfacegeo attributes in place with newly select target.

juzen2003 committed:
Update get_triggered_tables to push in surfacegeo target table even if the
it has 0 result. This will make sure newly selected surfacegeo target has
related menu item displayed in search sidebar.
- In views.py (line 1580-1582), push in table to triggered_tables even if
  the newly selected surfacegoe target has 0 result.
- In menu.js, clean up the indentation.

juzen2003 committed:
Clean up commented out codes and console.log in search.js and widgets.js.

juzen2003 committed:
Update the unit record for surfacegeo target related widgets so that after
switching to a new target and changing unit, the values can be properly
converted.
- In widgets.js (line 157-158), update record in opus.currentUnitBySlug
  when a new surfacegeo target is selected.
- In widgets.js (line 204-217), create updateSURFACEGEODataObj to update
  surfacegeo info in data object (opus.selections, opus.extras, and
  opus.currentUnitBySlug) with newly select target.

juzen2003 committed:
Close all related SURFACEGEO widgets when the surfacegeometrytargetname
widget is closed.
- In views.py (line 1581), update comments.
- In hash.js (line 643-645), update comments.
- In utils.js (line 114), add comments.
- In import_util.py (line 266), update comments.
- In widgets.js (line 220, 161), update comments.
- In widgets.js (line 111), add a if block to check if the singlechoice
  is a surfacegeo target input, if so, we update surfacegeo widgets in place.
- In widgets.js (line 232-247), create closeAllSURFACEGEOWidgets to close
  all SURFACEGEO related widgets.
- In widgets.js (line 101-103), if the closing widget is the surfacegeometrytargetname
  widget, we will also close all the related SURFACEGEO widgets.
- In widgets.js (line 44-48), move isRemovingEmptyWidget from closeWidget
  to o_widgets. This flag is used to determined if we are closing any
  widget with empty input. In a series of closing widget actions, if there
  is one input in a widget with a value, we will make sure a search is
  performed after updating URL. If none of inputs has value, we don't need
  to perform a search.
- In widget.js/opus.js (3 places), remove delete opus.normalizeInputForAllFieldsInProgress[opus.allSlug]
  if normalizedData.reqno < o_search.lastSlugNormalizeRequestNo. This
  solve the page reloading issue if we keep closing lots of widget at one
  time.

juzen2003 committed:
Expand intended target and surfacegeo input group on permalink.
- In views.py (line 293), add mult group label as the data-group attributes
  of ul element (.mult_group).
- In widgets.js (line 838-852), create expandInputGroup to loop through
  targetInputs and expand the group with selected inputs.
- In widgets.js (line 814-825), modify customWidgetBehaviors to call
  expandInputGroup to expand the selected inputs group on permalink.

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into 872_ui_change_for_surface_geometry_searching

juzen2003 committed:
Restore the original behavior that when a planet is selected, the corresponding
input groups in intended target and surfacegeo target are expanded.
- In widgets.js (line 882-891), create expandInputGroupBySelectedPlanet
  to organize the old code to expand the corresponding input groups based
  on the selected planet.
- In widgets.js (line 816-820, 824-831, 840-847), restore old code with
  expandInputGroupBySelectedPlanet.

juzen2003 committed:
Pop out a confirmation modal when closing surfacegeometrytargetname widget
if there are other SURFACEGEO related widgets open. If there are no other
SURFACEGEO related widgets open, there will be no confirmation modal.
- In base.html (line 488-498), add a confirmation modal #op-close-surfacegeo-widgets.
  It will be popped out when closing surfacegeometrytargetname widget
  if there are other SURFACEGEO related widgets open.
- In opus.js (line 837-839), call o_widgets.closeAllSURFACEGEOWidgets()
  to close all surfacegeo related widgets when submit button is hit.
- In widgets.js (line 249-258), re-organize the code and create
  closeAndRemoveWidgetFromDOM to close widget and remove widget elements
  from DOM.
- In widgets.js (line 165-173), update metadata selections (opus.prefs.widgets)
  when surfacegeo target is changed.

juzen2003 committed:
Update metadata selections when surfacegeotarget is changed in search tab
and there are surfacegeo metadata items (related to old target) selected.
- In browse.js (line 1616-1618), in getDataURL, get the hash string from
  getFullHashStr instead of getHash because we want the updated version
  of cols= even if the main URL hasn't been updated yet.
- In browse.js (line 1896-1898), in loadData, increment opus.lastLoadDataRequestNo[view].
  Previously we didn't increment this reqno for dataimages.json api call.
- In widgets.js (line 170-173), update metadata selections when surfacegeotarget
  is changed and there are surfacegeo metadata items (related to old target)
  selected.

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into 872_ui_change_for_surface_geometry_searching

juzen2003 committed:
Update gettingStarted guide for surfacegeometry field, update the title
and messages in #op-close-surfacegeo-widgets modal, and add the new
surfacegeo metadata items at the end of opus.prefs.col.
- In gettingstarted.html (line 121-133), add new info for surfacegeometry
  field.
- In base.html (line 490-492), update the title and messages of
  #op-close-surfacegeo-widgets modal.
- In widgets.js (line 176-177), append the new surfacegeometry metadata
  selections to the end of opus.prefs.col.

juzen2003 committed:
Update the error message when and error occurs at widget closing event.
- In widdgets.js (line 262), update the error message.

juzen2003 committed:
When there is any value in SURFACEGEO related widgets, in surfacegeometrytargetname
, we will have "--" next to non-selected inputs and green hint number next
to the selected input. When there is no value in SURFACEGEO related widgets,
in surfacegeometrytargetname, we will have green hints (numbers) next to all inputs.
- In opus.js (line 375-379), update the logic to have "--" next to non-selected
  inputs and green hint number next to the selected input in surfacegeometrytargetname
  when there are values in SURFACEGEO related widgets.
- In search.js (line 1155-1181), add a parameter hideHintsForSurfacegeoTarget
  to getValidMults, if it's true, have "--" next to non-selected inputs
  and green hint number next to the selected input in surfacegeometrytargetname.

juzen2003 committed:
Fix the comment in opus.js (line 373).

juzen2003 committed:
Update the get hints logic in getWidget, this will make sure when user
switches back to search tab, the hints are properly displayed for the
surfacegeometrytargetname widget.
- In widgets.js (line 1291-1295), update the logic for getting hints in
  getWidget.

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into 872_ui_change_for_surface_geometry_searching

juzen2003 committed:
Modify SURFACEGEO slugs to have "_" as the delimiter to separate out
surfacegeo target and rest of the slug.
- In search/views.py (line 455-478), ui/views.py (line 595-619), update
  the logic to parse clause_num_str and slug for SURFACEGEO slugs because
  they might have multiple "_".
- In widgets.js (line 111-112), add "_" to the end of oldTargetStr and
  newTargetStr
- In obs_surface_geometry_target.json, modify pi_slug to have "_" as the
  delimiter, and update pi_old_slug with the original pi_slug value.

juzen2003 committed:
Hide "Surface Geometry Target Name" in the "Current Search Terms" part
of select metadata.
- In views.py (line 1485-1488), don't store surfacegeometrytargetname
  into menu_data. This will make sure "Surface Geometry Target Name" field
  will never be shown in the "Current Search Terms" part of select metadata.

juzen2003 committed:
Fixed the issue that radio buttons get deselected when dragging (ui-sortable)
a widget with radio buttons (surfacegeometrytargetname).
- In widgets.js (line 224-237), create recordRadioButtonStatus to record
  radio buttons checked status in data-checked of in corresponding inputs.
- In widgets.js (line 63-69), in ui-sortable stop event handler, restore
  radio buttons checked status.
- In widgets.js (line 193), call recordRadioButtonStatus whenever a radio
  button change event happened.

juzen2003 committed:
Fixed the issue that add(+) input icon in SURFACEGEO will not work after
surfacegeometry target is changed.
- In widgets.js (line 340-341), access data attributes by attr(), because
  we update the DOM attributes by attr() and the updated data will be
  displayed in html as well. Therefore we will use attr() to access the
  updated data. Using .data() will not work for accessing the updated data.

juzen2003 committed:
Update the logic to parse for clause_num. Now we could have slug with
multiple "_", so we will always get the last index of "_" in a slug, and
parse it. If it's a positive int, we will use it as an valid clause number,
else, we will leave the slug alone, and the invalid slug will be catched
later on as the unknown slug.
- In search/views.py (line 455-467), update the logic to parse for clause_num.
- In ui/views.py (line 595-606), update the logic to parse for clause_num.
- In hash.js (line 223-225), correct the logic in convertSlugForSorting
  when getting trailing counter string. This will fix the duplicated
  qtype & unit slugs in URL when surfacegeo target name is changed.
- In widgets.js (line 245-246), add "_" to the end of oldTargetStr &
  newTargetStr. This will make sure related strings are replace properly.

juzen2003 committed:
Fixed django tests, including the returned error messages and "_" in the
expected json for SURFACEGEO in test_ui_api.py and
api_metadata2_vg_iss_2_s_c4360845_default_json.json.

juzen2003 committed:
Add some django tests for SURFACEGEO.

juzen2003 committed:
Update the regex pattern for slug in log analyzer.
- In slug.py (line 129), add "$" to the end of the regex pattern to
  properly catch the clause number & correct slug when "_" exists.

rfrenchseti commented on opus/application/apps/search/views.py:
```diff
@@ -452,16 +452,19 @@ def url_to_search_params(request_get, allow_errors=False, return_slugs=False,
-                          'number "%s"', orig_slug)
-                return None, None
+                # If clause_num is not a positive integer, leave the slug as is.
+                # If the slug is unknown, it will be catched later as the
```
catched => caught and "as the" => "as an"

rfrenchseti commented on opus/application/apps/ui/views.py:
```diff
@@ -591,17 +592,18 @@ def _sort_func(key):
-                msg_list.append(msg)
-                continue
+                # If clause_num is not a positive integer, leave the slug as is.
+                # If the slug is unknown, it will be catched later as the
```
catched => caught and "as the" => "as an"

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1891,6 +1893,9 @@ var o_browse = {
         // then after the load is complete, instead of hiding the galleryView slide, update the metadata.
         let updateMetadataBox = $("#op-select-metadata").hasClass("show") && $("#galleryView").hasClass("show") ;
 
+        // Note: Increment the reqno here instead of getDataURL because infiniteScroll path also use getDataURL,
```
also use => also uses

rfrenchseti replied:
reqno for infiniteScroll path => reqno for the infiniteScroll URL

rfrenchseti commented on opus/application/static_media/js/opus.js:
```diff
@@ -372,7 +372,11 @@ var opus = {
         // Finally, update all the hints
         $.each(opus.prefs.widgets, function(index, slug) {
-            o_search.getHinting(slug);
+            if (slug === "surfacegeometrytargetname" && !o_search.isAllSURFACEGEOSelectionsEmpty()) {
```
I think the grammar would be better as "areAll..."

rfrenchseti commented on opus/application/static_media/js/search.js:
```diff
@@ -556,6 +559,20 @@ var o_search = {
+    isAllSURFACEGEOSelectionsEmpty: function() {
+        let isInputSetEmpty = true;
+        for (const slug in opus.selections) {
+            if (slug.match(/^SURFACEGEO/)) {
```
slug.startsWith?

rfrenchseti commented on opus/application/static_media/js/search.js:
```diff
@@ -556,6 +559,20 @@ var o_search = {
+        let isInputSetEmpty = true;
+        for (const slug in opus.selections) {
+            if (slug.match(/^SURFACEGEO/)) {
+                isInputSetEmpty = o_search.isSlugSelectionsEmpty(opus.selections[slug], isInputSetEmpty);
```
is => are

rfrenchseti commented on opus/application/static_media/js/search.js:
```diff
@@ -1112,7 +1129,7 @@ var o_search = {
     },
 
-    getValidMults: function(slug) {
+    getValidMults: function(slug, hideHintsForSurfacegeoTarget=false) {
```
This doesn't actually test for surfacegeotarget and works for any widget with radio buttons, right? So maybe it should be called "hideHintsForNonSelectedRadioButtons" or something like that.

rfrenchseti commented on opus/application/static_media/js/widgets.js:
```diff
@@ -84,27 +99,98 @@ var o_widgets = {
+
+                // Update widgets array in opus.prefs
+                $.each(opus.prefs.widgets, function(widgetIdx, widget) {
+                    if (widget.match(oldTargetStr)) {
```
startsWith?

rfrenchseti commented on opus/application/static_media/js/widgets.js:
```diff
@@ -84,27 +99,98 @@ var o_widgets = {
+                let currentMetadataSelections = opus.prefs.cols.slice();
+                let newSurfacegeoMetadataSelections = [];
+                for (const col of currentMetadataSelections) {
+                    if (col.match(oldTargetStr)) {
```
startsWith?

rfrenchseti commented on opus/application/static_media/js/widgets.js:
```diff
@@ -135,6 +221,72 @@ var o_widgets = {
+         * NOTE: the passed in dataObj will be modified.
+         */
+        for (const oldTargetSlug in dataObj) {
+            if (oldTargetSlug.match(oldTargetStr)) {
```
startsWith?

juzen2003 replied:
For this one, we will stay with .match because it's used to update the dataObj, including opus.extras, and there are "qtype-SURFACEGEO.*" and "unit-SURFACEGEO.*" that have "SURFACEGEO" in the middle of the string. So we will use .match, instead of .startsWith.

rfrenchseti commented on opus/application/test_api/test_ui_api.py:
```diff
@@ -2371,8 +2451,8 @@ def test__api_normalizeurl_real_5(self):
         "[test_ui_api.py] /__normalizeurl: real 6"
-        url = '/opus/__normalizeurl.json?SURFACEGEOplutocenterresolution2=5&SURFACEGEOplutophase1=160&surfacegeometrytargetname=Pluto&qtype-SURFACEGEOplutophase=any&cols=opusid,instrument,time1,SURFACEGEOplutoplanetographiclatitude1,SURFACEGEOplutoplanetographiclatitude2,SURFACEGEOplutoIAUwestlongitude1,SURFACEGEOplutoIAUwestlongitude2,SURFACEGEOplutocenterdistance,SURFACEGEOplutocenterresolution ...
-        expected = {"new_url": "surfacegeometrytargetname=Pluto&SURFACEGEOplutocenterresolution2=5&unit-SURFACEGEOplutocenterresolution=km/pixel&SURFACEGEOplutophase1=160&qtype-SURFACEGEOplutophase=any&unit-SURFACEGEOplutophase=degrees&cols=opusid,instrument,time1,SURFACEGEOplutoplanetographiclatitude1,SURFACEGEOplutoplanetographiclatitude2,SURFACEGEOplutoIAUwestlongitude1,SURFACEGEOplutoIAUwestl ...
+        url = '/opus/__normalizeurl.json?SURFACEGEOpluto_centerresolution2=5&SURFACEGEOpluto_phase1=160&surfacegeometrytargetname=Pluto&qtype-SURFACEGEOpluto_phase=any&cols=opusid,instrument,time1,SURFACEGEOpluto_planetographiclatitude1,SURFACEGEOpluto_planetographiclatitude2,SURFACEGEOpluto_IAUwestlongitude1,SURFACEGEOpluto_IAUwestlongitude2,SURFACEGEOpluto_centerdistance,SURFACEGEOpluto_centerr ...
```
I think you want to leave the old SURFACEGEO formats here (without the "_") in the URL so that we test normalizeurl's ability to switch from the old slug format to the new one.

rfrenchseti replied:
This wasn't entirely fixed so I went ahead and removed the remaining two "_" that shouldn't be there.

juzen2003 committed:
Fixed grammar in the comments of multiple files, update test cases in
test_ui_api.py. Make sure widget spinner will disappear when get hints
api is successfully returned. Also make sure metadata selector will be
re-rendered when surfacetarget name is changed and opus.prefs.cols is
updated.
- In search/views.py and ui/views.py, fixed grammar in the comments.
- In browse.js, fixed comments.
- In all files, rename "isAllSURFACEGEOSelectionsEmpty" to "areAllSURFACEGEOSelectionsEmpty".
- In all files, rename "isInputSetEmpty" to "areInputSetsEmpty".
- In all files, rename "isSlugSelectionsEmpty" to "areSlugSelectionsEmpty".
- In search.js, rename "hideHintsForSurfacegeoTarget" to "hideHintsForNonSelectedRadioButtons".
- In test_ui_api.py, update test case: /__normalizeurl: real 6.
- In widgets.js (line 182), set o_selectMetadata.rendered to false if
  opus.prefs.cols is updated when surfacegeo target name is changed. This
  will make sure metadata selector menu is properly updated.
- In search.js (line 1142), make sure widget spinner will go away when
  api is successfully returned.

juzen2003 committed:
Fixed the indentations at line 442,444,448 in search.js.

juzen2003 committed:
When metadata info is outdated, in o_selectMetadata.render function, we
will empty the old contents and display a spinner before the new contents
are loaded. We also detached the hide.bs.modal and show.bs.modal event handlers
for #op-select-metadata to make sure there is no multiple calls of api:
/opus/__forms/metadata_selector.html when "Select Metadata" is clicked.
- In selectMetadata.js (line 113-117), empty the old contents and display
  a spinner before the new contents are loaded for select metadata menu.
- In selectMetadata.js (line 131-137), remove #op-select-metadata's
  hide.bs.modal and show.bs.modal event handlers before attaching them
  once again. This will make sure /opus/__forms/metadata_selector.html
  gets called for only one time when "Select Metadata" is clicked.
- In selectMetadata.js (line 172-173), properly update perfect scrollbars
  in select metadata modal after they are initialized.
- In widgets.js (line 849), when widgetDrop happened (only happened in
  search tab), set o_selectMetadata.rendered to false and render select
  metadata menu contents when user clicks to enter browse tab instead just
  re-render select metadata menu contents right away. This will avoid the
  race condition of multiple /opus/__forms/metadata_selector.html calls
  when user slightly drags a widget while changing selected target and move
  forward to browse tab and click "Select Metadata" quickly.
